js/src/util.ts

```diff
@@ -0,0 +1,29 @@
+// Given a callback, runs the callback as a "finally" component. If
+// `callback` returns a Promise, `finallyF` is chained to the promise. Otherwise,
+// it is run afterwards synchronously.
+export function runFinally<R>(f: () => R, finallyF: () => void): R {
+  let runSyncCleanup = true;
+  try {
+    const ret = f();
+    // There's no "blessed" way to detect whether an object is a
+    // Promise-like object in javascript. So we use duck typing.
```

Comment:
In `framework.ts` we do:

```
 if (dataResult instanceof Promise) {
    data = await dataResult;
  } else {
    data = dataResult;
  }
```

does this not work?

and if not, should we change those places too?

Reply:
I was just scared off by the stack overflow crowd suggesting there are other kinds of "thenable" objects in other promise libraries: https://stackoverflow.com/questions/27746304/how-to-check-if-an-object-is-a-promise. Unclear if we need to really care about that.

Reply:
I'll just use `instanceof Promise` for consistency.